runtime/android/core/src/org/xwalk/core/XWalkActivity.java

```diff
@@ -5,16 +5,625 @@
+                            args[i] = mCoreWrapper.getBridgeObject(m.getInstance());
+                        } else {
+                            args[i] = m.invokeWithArguments();
+                        }
+                    }
+                }
+            }
+            method.invokeWithArguments();
+        }
+
+        mIsInitializingLibrary = false;
+        mIsXWalkReady = true;
+        onXWalkReady();
+    }
+
+    protected abstract void onXWalkReady();
+
+    protected void onXWalkStartupError(LibraryStatus status, Throwable e) {
+        Log.d("XWalkActivity", "XWalk startup error: " + status);
+        e.printStackTrace();
+
+        AlertDialog dialog = null;
+        if (status == LibraryStatus.NOT_FOUND) {
+            dialog = getStartupNotFoundDialog();
+        } else if (status == LibraryStatus.OLDER_VERSION) {
+            dialog = getStartupOlderVersionDialog();
+        } else if (status == LibraryStatus.NEWER_VERSION) {
+            dialog = getStartupNewerVersionDialog();
+        }
+        showDialog(dialog);
+    }
+
+    protected void onXWalkRuntimeError(LibraryStatus status, Throwable e) {
+        Log.d("XWalkActivity", "XWalk runtime error: " + status);
+        e.printStackTrace();
+
+        AlertDialog dialog = null;
+        if (status == LibraryStatus.OLDER_VERSION) {
+            dialog = getRuntimeOlderVersionDialog();
+        } else if (status == LibraryStatus.NEWER_VERSION) {
+            dialog = getRuntimeNewerVersionDialog();
+        }
+        showDialog(dialog);
+    }
+
+    protected boolean isXWalkReady() {
+        return mIsXWalkReady;
+    }
+
+    protected boolean isSharedMode() {
+        return mCoreWrapper != null && mCoreWrapper.isSharedMode();
+    }
+
+    protected boolean isShowingDialog() {
+        return mActiveDialog != null && mActiveDialog.isShowing();
+    }
+
+    protected int getSdkVersion() {
+        return XWalkSdkVersion.SDK_VERSION;
+    }
+
+    protected String getStringResource(String label) {
+        int resId = getResources().getIdentifier(label, "string", getPackageName());
+        if (resId == 0) return null;
+        return getString(resId);
+    }
+
+    protected ProgressDialog buildProgressDialog() {
+        return new ProgressDialog(this);
+    }
+
+    protected AlertDialog buildAlertDialog() {
+        return new AlertDialog.Builder(this).create();
+    }
+
+    protected void showDialog(Dialog dialog) {
+        mActiveDialog = dialog;
+        mActiveDialog.show();
+    }
+
+    protected void dismissDialog() {
+        mActiveDialog.dismiss();
+    }
+
+    private void getXWalkLibrary() {
+        try {
+            Intent intent = new Intent(Intent.ACTION_VIEW);
+            startActivity(intent.setData(Uri.parse(XWALK_CORE_MARKET_URL)));
+
+            showDialog(getMarketProgressDialog());
+        } catch (ActivityNotFoundException e) {
+            Log.d("XWalkActivity", "Market open failed");
+            downloadXWalkLibrary();
```

Comment:
Add a TODO to make it clear that need to at least edit the download url before final release.